### PR TITLE
Integration/bft testing

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

File: radixdlt/src/integration/java/com/radixdlt/consensus/ChainedBFTTest.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
+package com.radixdlt.consensus;
```
Comment (clstrfsck) [inline radixdlt/src/integration/java/com/radixdlt/consensus/ChainedBFTTest.java]: Missing file header comment.  Please set up IDE to do this automatically for new files.

File: radixdlt/src/main/java/com/radixdlt/CerberusModule.java [hunk 2]
```diff
@@ -53,15 +55,19 @@ protected void configure() {
...
-		bind(ProposerElection.class).to(DumbProposerElection.class);
-
 		bind(PacemakerRx.class).to(PacemakerImpl.class);
 		bind(Pacemaker.class).to(PacemakerImpl.class);
-
 		bind(SafetyRules.class).in(Scopes.SINGLETON);
 	}
 
+	@Provides
+	@Singleton
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/CerberusModule.java:63]: I think it's better to bind things using the binding DSL as much as possible, rather than using annotations.  If all the bindings are in one place it makes it much easier to see what is going on.

Ideally, remove the annotations from this method, then use 
```
bind(ProposerElection.class).toProvider(this::proposerElection).in(Scopes.SINGLETON);
```

Similarly for QuorumRequirements below.

File: radixdlt/src/main/java/com/radixdlt/consensus/ChainedBFT.java [hunk 3]
```diff
@@ -53,29 +72,27 @@ public ChainedBFT(
 		this.eventCoordinator = eventCoordinator;
 	}
 
-	public void start() {
-		this.pacemaker.start();
-
-		final Disposable timeoutDisposable = this.pacemaker.localTimeouts()
+	public Observable<Event> processEvents() {
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/ChainedBFT.java:83]: I think the word "process" just makes what this method with no Javadoc does very confusing.

File: radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java [hunk 4]
```diff
@@ -77,34 +77,42 @@ public EventCoordinator(
 		this.selfKey = Objects.requireNonNull(selfKey);
 	}
 
-	private void processNewView(View view) {
-		log.debug("Processing new view: " + view);
+	// TODO: Move to EUID.java
+	private String getShortName(EUID euid) {
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java:80]: Not convinced this is a good idea, given birthday paradox.

File: radixdlt/src/main/java/com/radixdlt/consensus/MempoolNetworkRx.java [hunk 5]
```diff
@@ -15,7 +15,7 @@
  * language governing permissions and limitations under the License.
  */
 
-package com.radixdlt.network;
+package com.radixdlt.consensus;
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/MempoolNetworkRx.java]: Why move this into the consensus package?  Either network or mempool package would be a better choice IMO.

If you are going to move it into the consensus package, then the DI setup also needs to move into the consensus module.

File: radixdlt/src/main/java/com/radixdlt/consensus/MempoolNetworkTx.java [hunk 6]
```diff
@@ -15,7 +15,7 @@
  * language governing permissions and limitations under the License.
  */
 
-package com.radixdlt.network;
+package com.radixdlt.consensus;
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/MempoolNetworkTx.java]: Similarly to above, don't understand the rationale for this move.

File: radixdlt/src/main/java/com/radixdlt/consensus/Vote.java [hunk 7]
```diff
@@ -86,6 +86,11 @@ private void setSerializerAuthor(byte[] author) throws CryptoException {
 		this.author = (author == null) ? null : new ECPublicKey(author);
 	}
 
+	@Override
+	public String toString() {
+		return String.format("Vote{author=%s view=%s}", author.getUID().toString().substring(0, 6), vertexMetadata.getView());
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/Vote.java]: Would prefer to see getClass().getSimpleName() rather than hardcoded classname here.

File: radixdlt/src/main/java/com/radixdlt/consensus/liveness/RotatingLeaders.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
+package com.radixdlt.consensus.liveness;
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/liveness/RotatingLeaders.java:18]: Missing header comment.

File: radixdlt/src/main/java/com/radixdlt/network/NetworkModule.java [hunk 9]
```diff
@@ -19,43 +19,26 @@
...
+import com.radixdlt.consensus.EventCoordinatorNetworkSender;
+import com.radixdlt.consensus.MempoolNetworkRx;
+import com.radixdlt.consensus.MempoolNetworkTx;
 
 public class NetworkModule extends AbstractModule {
 
-	private final boolean useNetwork;
-
 	public NetworkModule() {
-		this(true);
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/network/NetworkModule.java:28]: Unnecessary constructor can be removed altogether.

**Review by clstrfsck — CHANGES_REQUESTED** [review]
There are some things that need sorting out:

1. There is a checkstyle warning, preventing travis from doing it's thing here: https://travis-ci.com/github/radixdlt/radixdlt-core/jobs/301052735#L333
2. getVertexStore(...) in CerberusModule no longer throws RadixEngineException, but can't comment on line because it didn't change
3. Code coverage at ~32% is too low.

File: radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinatorNetworkSender.java [hunk 10]
```diff
@@ -15,17 +15,15 @@
...
+import com.radixdlt.common.EUID;
 
 /**
  * Interface for Event Coordinator to send things through a network
  */
 public interface EventCoordinatorNetworkSender {
 	void broadcastProposal(Vertex vertex);
-	void sendNewView(NewView newView);
-	void sendVote(Vote vote);
+	void sendNewView(NewView newView, EUID newViewLeader);
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinatorNetworkSender.java:27]: Do we want to identify nodes by their EUID? Pub keys seem more straightforward to me

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java [hunk 11]
```diff
@@ -84,13 +90,23 @@ public void insertVertex(Vertex vertex) throws VertexInsertionException {
...
+		if (tipVertex == null) {
 			throw new IllegalStateException("Committing a vertex which was never inserted: " + vertexId);
 		}
+		Vertex vertex = tipVertex;
+		while (vertex != null && !committedVertices.containsKey(vertex.getId())) {
+			committedVertices.put(vertexId, vertex);
+			vertex = vertices.get(vertex.getParentId());
+		}
+
+		lastCommittedVertex.onNext(tipVertex);
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java:103]: This is where we would/will also be doing some cleanup of no longer committable branches, right?

File: radixdlt/src/main/java/com/radixdlt/consensus/liveness/Pacemaker.java [hunk 12]
```diff
@@ -27,7 +27,13 @@
...
-	Optional<View> processRemoteNewView(NewView newView);
+
+	/**
+	 * Notifies pacemaker that the local timeout is being processed.
+	 * @param view the view to timeout
+	 * @return optional with view, if the pacemaker still agrees with the current view to timeout
+	 */
+	Optional<View> processLocalTimeout(View view);
 	Optional<View> processQC(View view);
+	Optional<View> processRemoteNewView(NewView newView);
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/liveness/Pacemaker.java]: I think it's time to add documentation to the other methods here as well

File: radixdlt/src/main/java/com/radixdlt/consensus/liveness/PacemakerImpl.java [hunk 13]
```diff
@@ -36,19 +36,23 @@
  * Overly simplistic pacemaker
  */
 public final class PacemakerImpl implements Pacemaker, PacemakerRx {
-	static final int TIMEOUT_MILLISECONDS = 500;
+	static final int TIMEOUT_MILLISECONDS = 1000;
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/liveness/PacemakerImpl.java:39]: Any specific reason for the increase? 
Adding a comment in source to explain the magic number would be sufficient for now

File: radixdlt/src/main/java/com/radixdlt/consensus/liveness/ProposerElection.java [hunk 14]
```diff
@@ -31,4 +31,11 @@
 	 * @return Whether the node is a valid proposer
 	 */
 	boolean isValidProposer(EUID nid, View view);
+
+	/**
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/liveness/ProposerElection.java:35]: Do we want to think about primary/secondary proposers? 
If we only want to deal with primary proposers, having both isValidProposer and getProposer is redundant.

File: radixdlt/src/main/java/com/radixdlt/consensus/liveness/SingleLeader.java [hunk 15]
```diff
@@ -19,11 +19,23 @@
 
 import com.radixdlt.common.EUID;
 import com.radixdlt.consensus.View;
+import java.util.Objects;
+
+public final class SingleLeader implements ProposerElection {
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/liveness/SingleLeader.java]: Might be more aptly named "SingleConstantLeader" or just "Dictatorship"

**Review by flotothemoon — COMMENTED** [review]
Looks like a good start but I need to take a deeper look

File: radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java (hunk 4 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java:80]: this is just for practical purposes unless there's a better idea in IDing nodes within a log?

File: radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinatorNetworkSender.java (hunk 10 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinatorNetworkSender.java:27]: Yes, probably. But perhaps for another task. Will add to jira.

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java (hunk 11 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java:103]: yes

File: radixdlt/src/main/java/com/radixdlt/consensus/MempoolNetworkRx.java (hunk 5 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/MempoolNetworkRx.java]: agree, moved back

File: radixdlt/src/main/java/com/radixdlt/consensus/MempoolNetworkTx.java (hunk 6 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/MempoolNetworkTx.java]: agree, moved back

File: radixdlt/src/main/java/com/radixdlt/CerberusModule.java (hunk 2 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/CerberusModule.java:63]: Not sure how this works given the specific injection (selfKey) needed?

File: radixdlt/src/main/java/com/radixdlt/consensus/MempoolNetworkRx.java (hunk 5 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/MempoolNetworkRx.java]: agree, changed.

File: radixdlt/src/main/java/com/radixdlt/consensus/MempoolNetworkTx.java (hunk 6 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/MempoolNetworkTx.java]: agree, changed.

File: radixdlt/src/integration/java/com/radixdlt/consensus/MultiNodeConsensusTest.java [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+		);
+	}
+
+	private ChainedBFT createBFTInstance(
+		ECKeyPair key,
+		ProposerElection proposerElection,
+		QuorumRequirements quorumRequirements,
+		VertexStore vertexStore
+	) {
+		Mempool mempool = mock(Mempool.class);
```
Comment (flotothemoon) [inline radixdlt/src/integration/java/com/radixdlt/consensus/MultiNodeConsensusTest.java:82]: Why are we mocking mempool but nothing else?

File: radixdlt/src/integration/java/com/radixdlt/consensus/MultiNodeConsensusTest.java [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+				VertexStore vertexStore = new VertexStore(genesisVertex, genesisQC, radixEngine);
+				TestObserver<Vertex> testObserver = TestObserver.create();
+				vertexStore.lastCommittedVertex().subscribe(testObserver);
+				ChainedBFT chainedBFT = createBFTInstance(e, proposerElection, quorumRequirements, vertexStore);
+				return new Pair<>(testObserver, chainedBFT.processEvents());
+			})
+			.collect(Collectors.toList());
+
+		Observable.merge(bftEvents.stream().map(Pair::getSecond).collect(Collectors.toList()))
+			.subscribe();
```
Comment (flotothemoon) [inline radixdlt/src/integration/java/com/radixdlt/consensus/MultiNodeConsensusTest.java]: Could we subscribe directly inside the stream instead of passing `chainedBFT.processEvents()` to the "outside" through `Pairs`?

File: radixdlt/src/integration/java/com/radixdlt/consensus/MultiNodeConsensusTest.java [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+		);
+
+		return new ChainedBFT(
+			eventCoordinator,
+			testEventCoordinatorNetwork.getNetworkRx(key.getUID()),
+			pacemaker
+		);
+	}
+
+	private List<TestObserver<Vertex>> runBFT(
```
Comment (flotothemoon) [inline radixdlt/src/integration/java/com/radixdlt/consensus/MultiNodeConsensusTest.java:112]: Even though this is private it would be helpful to document what this returns (the stream of committed vertices)

File: radixdlt/src/integration/java/com/radixdlt/consensus/MultiNodeConsensusTest.java [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+		final QuorumRequirements quorumRequirements = WhitelistQuorum.from(nodes.stream().map(ECKeyPair::getPublicKey));
+		final ProposerElection proposerElection = new Dictatorship(nodes.get(0).getUID());
+		final List<TestObserver<Vertex>> committedListeners = runBFT(nodes, quorumRequirements, proposerElection);
+
+		final int commitCount = 3;
+		for (TestObserver<Vertex> committedListener : committedListeners) {
+			committedListener.awaitCount(commitCount);
+			for (int i = 0; i < commitCount; i++) {
+				final int id = i;
+				committedListener.assertValueAt(i, v -> v.getAtom().toString().equals(Integer.toHexString(id)));
```
Comment (flotothemoon) [inline radixdlt/src/integration/java/com/radixdlt/consensus/MultiNodeConsensusTest.java:144]: This seems to rely on the behaviour of `Atom.toString`, so would be cleaner to get and compare the desired property (id) directly.

Comment (flotothemoon) [reply radixdlt/src/integration/java/com/radixdlt/consensus/MultiNodeConsensusTest.java:144]: Actually, scratch that since Atom is mocked anyway. 
Please add a comment for any future reader who might be confused by this

File: radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java (hunk 4 above)
Comment (flotothemoon) [reply radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java:80]: pubKeys?

File: radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java [hunk 20]
```diff
@@ -77,34 +77,42 @@ public EventCoordinator(
...
-			log.warn(String.format("Ignoring confused vote %s for %s", vote.hashCode(), vote.getVertexMetadata().getView()));
-			return;
-		}
-
-		// accumulate votes into QCs in store
-		Optional<QuorumCertificate> potentialQc = this.pendingVotes.insertVote(vote);
-		if (potentialQc.isPresent()) {
-			QuorumCertificate qc = potentialQc.get();
-			processQC(qc);
+	private void proceedToView(View nextView) {
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java:105]: There's some odd re-ordering going on here, are these changes intentional? e.g. moving processVote

File: radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java [hunk 21]
```diff
@@ -77,34 +77,42 @@ public EventCoordinator(
...
+		try {
+			// TODO make signing more robust by including author in signed hash
+			ECDSASignature signature = this.selfKey.sign(Hash.hash256(Longs.toByteArray(nextView.number())));
+			NewView newView = new NewView(selfKey.getPublicKey(), nextView, this.vertexStore.getHighestQC(), signature);
+			EUID nextLeader = this.proposerElection.getProposer(nextView);
+			log.info(this.getShortName() + ": Sending NewView to " + this.getShortName(nextLeader) + ": " + newView);
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java]: Might be cleaner to use good ol' String.format here

File: radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java [hunk 22]
```diff
@@ -115,53 +123,77 @@ private void processQC(QuorumCertificate qc) {
...
-			// TODO make signing more robust by including author in signed hash
-			ECDSASignature signature = this.selfKey.sign(Hash.hash256(Longs.toByteArray(view.next().number())));
-			this.networkSender.sendNewView(new NewView(selfKey.getPublicKey(), view.next(), signature));
-		} catch (CryptoException e) {
-			throw new IllegalStateException("Failed to sign new view at " + view, e);
+		// accumulate votes into QCs in store
+		Optional<QuorumCertificate> potentialQc = this.pendingVotes.insertVote(vote);
+		if (potentialQc.isPresent()) {
+			QuorumCertificate qc = potentialQc.get();
+			log.info(this.getShortName() + ": Creating QC: " + qc);
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java]: "Formed" instead of "Creating" would be more accurate (nitpicking)

File: radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinatorNetworkSender.java (hunk 10 above)
Comment (flotothemoon) [reply radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinatorNetworkSender.java:27]: Fair enough, that works

File: radixdlt/src/main/java/com/radixdlt/consensus/NewView.java [hunk 23]
```diff
@@ -56,12 +60,17 @@
...
 
-	public NewView(ECPublicKey author, View view, ECDSASignature signature) {
+	public NewView(ECPublicKey author, View view, QuorumCertificate qc, ECDSASignature signature) {
 		this.author = Objects.requireNonNull(author);
 		this.view = Objects.requireNonNull(view);
+		this.qc = qc;
 		this.signature = signature;
 	}
 
+	public QuorumCertificate getQc() {
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/NewView.java]: `getQC` should be capitalised according to our style

File: radixdlt/src/main/java/com/radixdlt/consensus/liveness/Dictatorship.java [hunk 24]
```diff
@@ -19,11 +19,23 @@
 
 import com.radixdlt.common.EUID;
 import com.radixdlt.consensus.View;
+import java.util.Objects;
+
+public final class Dictatorship implements ProposerElection {
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/liveness/Dictatorship.java:27]: Document me

File: radixdlt/src/integration/java/com/radixdlt/consensus/MultiNodeConsensusTest.java (hunk 16 above)
Comment (talekhinezh) [reply radixdlt/src/integration/java/com/radixdlt/consensus/MultiNodeConsensusTest.java:82]: mempool is the only module that sort exists outside of BFT consensus space, so can mock it so we don't need to worry about client interactions with system.

File: radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java (hunk 4 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java:80]: the issue is that the string/name of the ID is long and so painful to use when debugging logs.

File: radixdlt/src/main/java/com/radixdlt/consensus/liveness/Pacemaker.java [hunk 25]
```diff
@@ -26,8 +26,31 @@
  * Interface for an event coordinator to manage the pacemaker
  */
 public interface Pacemaker {
+
+	/**
+	 * Retrieves the local current view the pacemaker is at
+	 * @return view of the pacemaker
+	 */
 	View getCurrentView();
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/liveness/Pacemaker.java:34]: Seems like this could be an observable too (like last committed vertex) for consistency, but not sure if necessary

File: radixdlt/src/main/java/com/radixdlt/consensus/liveness/RotatingLeaders.java [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+package com.radixdlt.consensus.liveness;
+
+import com.google.common.collect.ImmutableList;
+import com.radixdlt.common.EUID;
+import com.radixdlt.consensus.View;
+import java.util.Objects;
+
+public final class RotatingLeaders implements ProposerElection {
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/liveness/RotatingLeaders.java:28]: Document me too

File: radixdlt/src/main/java/com/radixdlt/network/TestEventCoordinatorNetwork.java [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+/**
+ * Overly simplistic network implementation that just sends messages to itself.
+ */
+public class TestEventCoordinatorNetwork {
+	private static final int LOOPBACK_DELAY = 50;
+	private final PublishSubject<Vertex> proposals;
+	private final PublishSubject<Map.Entry<NewView, EUID>> newViews;
+	private final PublishSubject<Map.Entry<Vote, EUID>> votes;
+	private final ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+	private final Set<EUID> sendingDisabled = new HashSet<>();
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/network/TestEventCoordinatorNetwork.java]: Doesn't this need to be thread-safe?

File: radixdlt/src/main/java/com/radixdlt/consensus/liveness/Pacemaker.java (hunk 25 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/liveness/Pacemaker.java:34]: Don't think so, this has to be synchronous with the rest of the operations. Would be much more difficult if implemented as async

File: radixdlt/src/main/java/com/radixdlt/network/TestEventCoordinatorNetwork.java (hunk 27 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/network/TestEventCoordinatorNetwork.java]: good catch

File: radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java (hunk 4 above)
Comment (clstrfsck) [reply radixdlt/src/main/java/com/radixdlt/consensus/EventCoordinator.java:80]: OK, I concede. Can we leave this here, rather than migrating to EUID as comment would suggest, as this really needs to remain a node specific thing.

File: radixdlt/src/main/java/com/radixdlt/network/TestEventCoordinatorNetwork.java [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+						proposals.onNext(vertex);
+					}, LOOPBACK_DELAY, TimeUnit.MILLISECONDS);
+				}
+			}
+
+			@Override
+			public void sendNewView(NewView newView, EUID newViewLeader) {
+				if (!sendingDisabled.contains(euid)) {
+					executorService.schedule(() -> {
+						newViews.onNext(new SimpleEntry<>(newView, newViewLeader));
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/network/TestEventCoordinatorNetwork.java]: Why is this not using Pair?

File: radixdlt/src/integration/java/com/radixdlt/consensus/MultiNodeConsensusTest.java (hunk 17 above)
Comment (talekhinezh) [reply radixdlt/src/integration/java/com/radixdlt/consensus/MultiNodeConsensusTest.java]: good idea

**Review by clstrfsck — APPROVED** [review]
Question from @flotothemoon about thread-safety with HashMap needs sorting, otherwise OK.
